Distributed tracing 2.9 (#62410)
openshift/openshift-docs: modules/distr-tracing-otel-migrating-from-jaeger-with-sidecars.adoc

  namespace: <otel-collector-namespace>
spec:
  mode: sidecar
  config: |
    receivers:
      jaeger:
        protocols:
          grpc:
          thrift_binary:
          thrift_compact:
          thrift_http:
    processors:
      batch:
      memory_limiter:
        check_interval: 1s
        limit_percentage: 50
        spike_limit_percentage: 30
      resourcedetection:
        detectors: [openshift]
        timeout: 2s
    exporters:
      otlp:
        endpoint: "tempo-<example>-gateway:8090" <1>
        tls:
          insecure: true
    service:
      pipelines:
        traces:
          receivers: [jaeger]
          processors: [memory_limiter, resourcedetection, batch]
          exporters: [otlp]
----
<1> This endpoint points to the Gateway of a TempoStack instance deployed by using the `<example>` {TempoOperator}.

. Create a service account for running your application.
+
[source,yaml]
----
apiVersion: v1
kind: ServiceAccount
metadata:
  name: otel-collector-sidecar
----

. Create a cluster role for the permissions needed by some processors.
+
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: otel-collector-sidecar
rules:

>   <1>

Reviewer: Possible to provide an example for the line with a `<1>`?

Author: Same reply as in https://github.com/openshift/openshift-docs/pull/62410#discussion_r1321079006.

- apiGroups: ["config.openshift.io"]
  resources: ["infrastructures", "infrastructures/status"]
  verbs: ["get", "watch", "list"]
----
<1> The `resourcedetectionprocessor` requires permissions for infrastructures and infrastructures/status.

> . Create a `ClusterRoleBinding` to set the permissions for the service account.

Reviewer: Per [docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#object-references):

>Use "object", "resource", "custom resource", "spec", etc. as appropriate after the object reference. This helps with clarity and readability.

Author: Adding this to my list. Same reply as in https://github.com/openshift/openshift-docs/pull/62410#discussion_r1321079006.

+
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: otel-collector-sidecar
subjects:
- kind: ServiceAccount
  name: otel-collector-deployment
  namespace: otel-collector-example
roleRef:
  kind: ClusterRole
  name: otel-collector
  apiGroup: rbac.authorization.k8s.io
----

. Deploy the OpenTelemetry Collector as a sidecar.

. Remove the injected Jaeger Agent from your application by removing the `"sidecar.jaegertracing.io/inject": "true"` annotation from your Deployment.

. Enable automatic injection of the OpenTelemetry sidecar by adding the `sidecar.opentelemetry.io/inject: "true"` annotation to the `.spec.template.metadata.annotations` field of your Deployment.

> . Use the created service account for the Deployment of your application to allow the processors to get the correct information and add it to your traces.

Reviewer: "deployment" or "`Deployment` object"
```suggestion
. Use the created service account for the deployment of your application to allow the processors to get the correct information and add it to your traces.
```

Author: Thank you. Applied manually. I also corrected this in all the other files.

// Module included in the following assemblies:
//
// * distr-tracing-otel-migrating.adoc

:_content-type: PROCEDURE
[id="distr-tracing-otel-migrating-from-jaeger_{context}"]
= (Temporary title) Migrating from {JaegerName} to {OTELName} with sidecars

.Prerequisites

> * Existing instrumentation of applications uses {JaegerName} for tracing.

Author: @iblancasa, I'm afraid we can't use the word `instrument` as a verb. How about some other wording? For example:
```suggestion
* {JaegerName} is used on the cluster.
```

* {OTELName} is installed.

.Procedure

> The Red Hat Red Hat OpenShift Distributed Tracing Data Collection Operator can inject sidecars into Deployment workloads.

Author: @iblancasa, since we have `:OTELName: Red Hat OpenShift distributed tracing data collection`, do you agree with the following suggestion?
```suggestion
The {OTELName} Operator can inject sidecars into Deployment workloads.
```

Configure OpenTelemetry Collector as sidecar

----
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: otel
  namespace: otel-collector-example
spec:
  mode: sidecar
  config: |
    receivers:
      jaeger:
        protocols:
          grpc:
          thrift_binary:
          thrift_compact:
          thrift_http:
    processors:
      batch:
      memory_limiter:
        check_interval: 1s
        limit_percentage: 50
        spike_limit_percentage: 30
      resourcedetection:
        detectors: [openshift]
        timeout: 2s
    exporters:
      otlp:

>         endpoint: "tempo-example-gateway:8090"

Author: ```suggestion
        endpoint: "tempo-example-gateway:8090" <1>
```

        tls:
          insecure: true
    service:
      pipelines:
        traces:
          receivers: [jaeger]
          processors: [memory_limiter, resourcedetection, batch]
          exporters: [otlp]

> ----

Author: Like this?
```suggestion
----
<1> This points to the Gateway of a Tempo instance deployed using the `example` Tempo Operator.
```

Create a Service Account for running your application:

----
apiVersion: v1
kind: ServiceAccount
metadata:
  name: otel-collector-sidecar
----

Create a ClusterRole for some permissions needed by some processors

----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: otel-collector-sidecar
rules:
# Permissions for infrastructures and infrastructures/status are needed for resourcedetectionprocessor
- apiGroups: ["config.openshift.io"]
  resources: ["infrastructures", "infrastructures/status"]
  verbs: ["get", "watch", "list"]
----

Create a ClusterRoleBinding to give those permissions to the Service Account

----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: otel-collector-sidecar
subjects:
- kind: ServiceAccount
  name: otel-collector-sidecar
roleRef:
  kind: ClusterRole
  name: otel-collector-sidecar
  apiGroup: rbac.authorization.k8s.io
----

Deploy the OpenTelemetry Collector as sidecar

----
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: otel
  namespace: otel-collector-example
spec:

>   mode: sidecar

Reviewer 2: Need to specify the created serviceAccount in the previous step.

```
spec:
  mode: sidecar
  serviceAccount: otel-collector-sidecar
  config: |
```

Reviewer 3: The serviceAccount needs to be set to the application where the sidecar is injected

        detectors: [openshift]
        timeout: 2s
    exporters:
      otlp:
        endpoint: "tempo-<example>-gateway:8090" <1>
        tls:
          insecure: true
    service:
      pipelines:
        traces:
          receivers: [jaeger]
          processors: [memory_limiter, resourcedetection, batch]
          exporters: [otlp]
----
<1> This points to the Gateway of a Tempo instance deployed by using the `<example>` Tempo Operator.

Create a Service Account for running your application:

----
apiVersion: v1
kind: ServiceAccount
metadata:
  name: otel-collector-sidecar
----

Create a ClusterRole for the permissions needed by some processors

----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: otel-collector-sidecar
rules:
# Permissions for infrastructures and infrastructures/status are needed for resourcedetectionprocessor
- apiGroups: ["config.openshift.io"]
  resources: ["infrastructures", "infrastructures/status"]
  verbs: ["get", "watch", "list"]
----

Create a ClusterRoleBinding to give those permissions to the Service Account

----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: otel-collector-sidecar

> subjects:

Reviewer 2: We need to specify the namespace for the subjects, otherwise the ClusterRoleBinding cannot be created.
`The ClusterRoleBinding "otel-collector-sidecar" is invalid: subjects[0].namespace: Required value`
Here is the example with namespace defined.
```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: otel-collector
subjects:
- kind: ServiceAccount
  name: otel-collector-deployment
  namespace: otel-collector-example
roleRef:
  kind: ClusterRole
  name: otel-collector
  apiGroup: rbac.authorization.k8s.io
```
